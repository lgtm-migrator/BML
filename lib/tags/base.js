const { BMLStream } = require('../BMLStream');
const { Context } = require('../context');
const { NotImplementedError } = require('../errors');
const { ensureTag } = require('../utils');

function prepareContext(contextData) {
  let context = new Context();
  if (contextData) {
    if (contextData instanceof Context) {
      context = contextData;
    } else {
      context.fill(contextData);
    }
  }
  return context;
}

class Tag {
  parse(data, contextData = {}) {
    const stream = new BMLStream(data);
    const context = prepareContext(contextData);
    return this._parse_(stream, context);
  }

  _parse_(_stream, _context) {
    throw new NotImplementedError(
      `_parse_ is not implemented in ${this.constructor.name}`
    );
  }

  pack(data, contextData = {}) {
    const stream = new BMLStream();
    const context = prepareContext(contextData);
    this._pack_(stream, data, context);
    return stream.finalize();
  }

  _pack_(_stream, _data, _context) {
    throw new NotImplementedError(
      `_parse_ is not implemented in ${this.constructor.name}`
    );
  }

  named(name) {
    this.name = name;
    return this;
  }

  toString() {
    return `${this.constructor.name} {
      ${this.name}
    }`;
  }
}

class CompositeTag extends Tag {
  constructor(...subTags) {
    super();
    this.subTags = subTags.map(ensureTag);
  }
}

class Adapter extends Tag {
  constructor(subTag) {
    super();
    this.tag = ensureTag(subTag);
  }

  _parse_(stream, context) {
    const data = this.tag._parse_(stream, context);
    return this._decode_(data, context);
  }

  _pack_(stream, data, context) {
    const encoded = this._encode_(data, context);
    this.tag._pack_(stream, encoded, context);
  }

  _decode_(_data, _context) {
    throw new NotImplementedError(
      `_decode_ is not implemented in ${this.constructor.name}`
    );
  }

  _encode_(_data, _context) {
    throw new NotImplementedError(
      `_encode_ is not implemented in ${this.constructor.name}`
    );
  }
}

function markTagFn(fn) {
  Object.defineProperty(fn, 'isTagFn', { get: () => true });
}

function tag(TagClass, args = []) {
  const tagFn = function tagFn(input) {
    let name;
    if (typeof input === 'string') {
      name = input; // calling tag with name as usual function
    } else if (Array.isArray(input) && input.raw) {
      [name] = input; // calling tag as template fn
    }
    function builderFn() {
      return new TagClass(...args).named(name);
    }
    markTagFn(builderFn);
    if (input === undefined) {
      return builderFn();
    }
    return builderFn;
  };
  Object.defineProperty(tagFn, 'name', {
    value: TagClass.name.toLowerCase(),
  });
  markTagFn(tagFn);
  return tagFn;
}

function tagWithParams(TagClass) {
  const tagParamFn = function tagParamFn(...args) {
    return tag(TagClass, args);
  };
  Object.defineProperty(tagParamFn, 'isTagFn', {
    get: () => {
      throw new Error(`params not passed for ${TagClass.name}`);
    },
  });
  return tagParamFn;
}

module.exports = {
  Tag,
  CompositeTag,
  Adapter,
  tag,
  tagWithParams,
};
