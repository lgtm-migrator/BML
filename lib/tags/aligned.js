const { Tag, tagWithParams } = require('./base');
const { ensureTag, functor } = require('../utils');

class Aligned extends Tag {
  constructor(subTag, modulus) {
    super();
    this.subTag = ensureTag(subTag);
    this.modulus = functor(modulus);
  }

  _parse_(stream, context) {
    const start_position = stream.tell();
    const value = this.subTag._parse_(stream, context);
    const end_position = stream.tell();
    const modulus = this.modulus(context);
    const pad = (end_position - start_position) % modulus;
    stream.seek(end_position + pad);
    return value;
  }

  _pack_(stream, data, context) {
    const start_position = stream.tell();
    this.subTag._pack_(stream, data, context);
    const end_position = stream.tell();
    const modulus = this.modulus(context);
    const pad = (end_position - start_position) % modulus;
    stream.seek(end_position + pad);
  }
}

module.exports = {
  aligned: tagWithParams(Aligned),
};
