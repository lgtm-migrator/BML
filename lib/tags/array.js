const { Tag, tagWithParams } = require('./base');
const { Context } = require('../Context');
const { functor, range, ensureTag } = require('../utils');

// Homogenous array of elements, similar to C# generic T[].
class ArrayTag extends Tag {
  constructor(tag, count) {
    super();
    this.tag = ensureTag(tag);
    this.count = functor(count);
  }

  _parse_(stream, context) {
    const count = this.count(context);
    const listContext = new Context(context);
    const data = [];
    for (const index of range(count)) {
      listContext.set('index', index);
      const subData = this.tag._parse_(stream, listContext);
      data.push(subData);
    }
    return data;
  }

  _pack_(stream, data, context) {
    const count = this.count(context);
    const listContext = new Context(context);
    for (const index of range(count)) {
      listContext.set('index', index);
      const subData = data[index];
      this.tag._pack_(stream, subData, listContext);
    }
  }

  _size_(context) {
    const listContext = new Context(context);
    return this.count(context) * this.tag._size_(listContext);
  }
}

module.exports = {
  array: tagWithParams(ArrayTag),
};
