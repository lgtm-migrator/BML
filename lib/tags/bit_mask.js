const { Tag, tagWithParams } = require('./base');
const { ensureTag } = require('../utils');

class BitMask extends Tag {
  constructor(subTag, bitMap) {
    super();
    this.subTag = ensureTag(subTag);
    this.bitMap = bitMap;
  }

  _parse_(stream, context) {
    const parsedValue = this.subTag._parse_(stream, context);
    const result = {};
    for (const key of Object.keys(this.bitMap)) {
      result[key] = Boolean(parsedValue & this.bitMap[key]);
    }
    return result;
  }

  _pack_(stream, data, context) {
    let value = 0;
    for (const key of Object.keys(this.bitMap)) {
      if (data[key]) {
        value |= this.bitMap[key];
      }
    }
    this.subTag._pack_(stream, value, context);
  }
}

module.exports = {
  bit_mask: tagWithParams(BitMask),
};
