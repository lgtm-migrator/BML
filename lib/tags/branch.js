const { Tag, tagWithParams } = require('./base');
const { functor, ensureTag } = require('../utils');

class Branch extends Tag {
  constructor(selector, cases, defaultTag) {
    super();
    this.selector = functor(selector);
    this.cases = cases;
    this.defaultTag = defaultTag;
  }

  _parse_(stream, context) {
    const caseKey = this.selector(context);
    let useCase = this.cases[caseKey];
    if (!useCase && this.defaultTag) {
      useCase = this.defaultTag;
    }
    if (useCase) {
      useCase = ensureTag(useCase);
      return useCase._parse_(stream, context);
    }
    return null;
  }

  _pack_(stream, data, context) {
    const caseKey = this.selector(context);
    let useCase = this.cases[caseKey];
    if (useCase) {
      useCase = ensureTag(useCase);
      useCase._pack_(stream, data, context);
    }
  }
}

module.exports = {
  branch: tagWithParams(Branch),
};
