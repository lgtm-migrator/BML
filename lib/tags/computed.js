const { Tag, tagWithParams } = require('./base');
const { functor } = require('../utils');

class Computed extends Tag {
  constructor(compute) {
    super();
    this.compute = functor(compute);
  }

  _parse_(_stream, context) {
    const computedValue = this.compute(context);
    if (this.name) {
      context.set(this.name, computedValue);
    }
    return computedValue;
  }

  _pack_(stream, _data, context) {
    this._parse_(stream, context);
  }

  _size_(_context) {
    return 0;
  }
}

const computed = tagWithParams(Computed);

module.exports = {
  computed,
};
