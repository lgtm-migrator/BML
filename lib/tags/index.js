const base = require('./base');
const primitives = require('./primitives');
const strings = require('./strings');
const enums = require('./enums');
const array = require('./array');
const sequence = require('./sequence');
const struct = require('./struct');
const constant = require('./constant');
const tap = require('./tap');
const when = require('./when');

module.exports = {
  ...base,
  ...primitives,
  ...strings,
  ...enums,
  ...array,
  ...sequence,
  ...struct,
  ...constant,
  ...tap,
  ...when,
};
