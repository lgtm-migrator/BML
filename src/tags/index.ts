export * from './tag';
export * from './adapter';
export * from './primitives';
export * from './strings';
export * from './enums';
export * from './array';
export * from './sequence';
export * from './struct';
export * from './constant';
export * from './tap';
export * from './when';
export * from './pointer';
export * from './seek';
export * from './skip';
export * from './pass';
export * from './branch';
export * from './computed';
export * from './repeat';
export * from './bitMask';
export * from './endian';
export * from './flag';
export * from './peek';
