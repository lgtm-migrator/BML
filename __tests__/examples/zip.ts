import * as path from 'path';
import { testFile } from '../helpers/tags';
import { zipFile } from '../../examples/zip.bml';

describe('zip files: ', () => {
  it('should parse', () => {
    testFile(path.join(__dirname, 'files/sample1.zip'), zipFile, 'zip');
  });
});
