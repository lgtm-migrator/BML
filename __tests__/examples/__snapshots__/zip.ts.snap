// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`zip files:  should parse: zip 1`] = `
Object {
  "sections": Array [
    Object {
      "body": Object {
        "body": Array [],
        "header": Object {
          "compressed_size": 0,
          "compression_method": "none",
          "crc32": 0,
          "extra_len": 0,
          "extras": Array [],
          "file_mod_date": 18782,
          "file_mod_time": 23810,
          "file_name": "folder/",
          "file_name_len": 7,
          "flags": 2,
          "uncompressed_size": 0,
          "version": 20,
        },
      },
      "section_type": "localFile",
    },
    Object {
      "body": Object {
        "body": Array [
          75,
          203,
          204,
          73,
          85,
          72,
          206,
          207,
          43,
          73,
          205,
          43,
          1,
          0,
        ],
        "header": Object {
          "compressed_size": 14,
          "compression_method": "deflated",
          "crc32": 3503491758,
          "extra_len": 0,
          "extras": Array [],
          "file_mod_date": 18782,
          "file_mod_time": 23804,
          "file_name": "folder/fileInFolder.txt",
          "file_name_len": 23,
          "flags": 2,
          "uncompressed_size": 12,
          "version": 20,
        },
      },
      "section_type": "localFile",
    },
    Object {
      "body": Object {
        "body": Array [
          75,
          203,
          204,
          73,
          85,
          72,
          206,
          207,
          43,
          73,
          205,
          43,
          1,
          0,
        ],
        "header": Object {
          "compressed_size": 14,
          "compression_method": "deflated",
          "crc32": 3503491758,
          "extra_len": 0,
          "extras": Array [],
          "file_mod_date": 18782,
          "file_mod_time": 23804,
          "file_name": "file.txt",
          "file_name_len": 8,
          "flags": 2,
          "uncompressed_size": 12,
          "version": 20,
        },
      },
      "section_type": "localFile",
    },
    Object {
      "body": Object {
        "comment": "",
        "comment_len": 0,
        "compressed_size": 0,
        "compression_method": "none",
        "crc32": 0,
        "disk_number_start": 0,
        "ext_file_attr": 16,
        "extra_len": 0,
        "extras": Array [],
        "file_name": "folder/",
        "file_name_len": 7,
        "flags": 2,
        "int_file_attr": 0,
        "last_mod_file_date": 18782,
        "last_mod_file_time": 23810,
        "local_header": Object {
          "body": Object {
            "body": Array [],
            "header": Object {
              "compressed_size": 0,
              "compression_method": "none",
              "crc32": 0,
              "extra_len": 0,
              "extras": Array [],
              "file_mod_date": 18782,
              "file_mod_time": 23810,
              "file_name": "folder/",
              "file_name_len": 7,
              "flags": 2,
              "uncompressed_size": 0,
              "version": 20,
            },
          },
          "section_type": "localFile",
        },
        "local_header_offset": 0,
        "uncompressed_size": 0,
        "version_made_by": 20,
        "version_needed_to_extract": 20,
      },
      "section_type": "centralDirEntry",
    },
    Object {
      "body": Object {
        "comment": "",
        "comment_len": 0,
        "compressed_size": 14,
        "compression_method": "deflated",
        "crc32": 3503491758,
        "disk_number_start": 0,
        "ext_file_attr": 32,
        "extra_len": 0,
        "extras": Array [],
        "file_name": "folder/fileInFolder.txt",
        "file_name_len": 23,
        "flags": 2,
        "int_file_attr": 1,
        "last_mod_file_date": 18782,
        "last_mod_file_time": 23804,
        "local_header": Object {
          "body": Object {
            "body": Array [
              75,
              203,
              204,
              73,
              85,
              72,
              206,
              207,
              43,
              73,
              205,
              43,
              1,
              0,
            ],
            "header": Object {
              "compressed_size": 14,
              "compression_method": "deflated",
              "crc32": 3503491758,
              "extra_len": 0,
              "extras": Array [],
              "file_mod_date": 18782,
              "file_mod_time": 23804,
              "file_name": "folder/fileInFolder.txt",
              "file_name_len": 23,
              "flags": 2,
              "uncompressed_size": 12,
              "version": 20,
            },
          },
          "section_type": "localFile",
        },
        "local_header_offset": 37,
        "uncompressed_size": 12,
        "version_made_by": 20,
        "version_needed_to_extract": 20,
      },
      "section_type": "centralDirEntry",
    },
    Object {
      "body": Object {
        "comment": "",
        "comment_len": 0,
        "compressed_size": 14,
        "compression_method": "deflated",
        "crc32": 3503491758,
        "disk_number_start": 0,
        "ext_file_attr": 32,
        "extra_len": 0,
        "extras": Array [],
        "file_name": "file.txt",
        "file_name_len": 8,
        "flags": 2,
        "int_file_attr": 1,
        "last_mod_file_date": 18782,
        "last_mod_file_time": 23804,
        "local_header": Object {
          "body": Object {
            "body": Array [
              75,
              203,
              204,
              73,
              85,
              72,
              206,
              207,
              43,
              73,
              205,
              43,
              1,
              0,
            ],
            "header": Object {
              "compressed_size": 14,
              "compression_method": "deflated",
              "crc32": 3503491758,
              "extra_len": 0,
              "extras": Array [],
              "file_mod_date": 18782,
              "file_mod_time": 23804,
              "file_name": "file.txt",
              "file_name_len": 8,
              "flags": 2,
              "uncompressed_size": 12,
              "version": 20,
            },
          },
          "section_type": "localFile",
        },
        "local_header_offset": 104,
        "uncompressed_size": 12,
        "version_made_by": 20,
        "version_needed_to_extract": 20,
      },
      "section_type": "centralDirEntry",
    },
    Object {
      "body": Object {
        "central_dir_offset": 156,
        "central_dir_size": 176,
        "comment": "",
        "comment_len": 0,
        "disk_of_central_dir": 0,
        "disk_of_end_of_central_dir": 0,
        "qty_central_dir_entries_on_disk": 3,
        "qty_central_dir_entries_total": 3,
      },
      "section_type": "endOfCentralDir",
    },
  ],
}
`;
